src/back/msl/writer.rs

```diff
@@ -1284,11 +1284,25 @@ impl<W: Write> Writer<W> {
                     }
                     let fun_name = &self.names[&NameKey::Function(function)];
                     write!(self.out, "{}(", fun_name)?;
+
+                    let function_arguments =
+                        &context.expression.module.functions[function].arguments;
+
                     // first, write down the actual arguments
-                    for (i, &handle) in arguments.iter().enumerate() {
+                    for (i, (&handle, argument)) in
+                        arguments.iter().zip(function_arguments).enumerate()
+                    {
                         if i != 0 {
                             write!(self.out, ", ")?;
                         }
+
+                        let ty = &context.expression.module.types[argument.ty];
+
+                        // If the function expects a pointer as an argument, take a reference.
```

Comment:
Is there a reason for this to be handled specifically for function parameters? I'd expect it to be handled more generally by `put_expression` instead.

Reply:
That probably would be better, but I'm not sure how that should be done. In the trace I just have

```
[TRACE naga::back::msl::writer] statement[1] Call { function: [1], arguments: [[15], [16], [17]], result: Some([18]) }
[TRACE naga::back::msl::writer] expression [15] = LocalVariable([1])
[TRACE naga::back::msl::writer] expression [16] = LocalVariable([2])
[TRACE naga::back::msl::writer] expression [17] = LocalVariable([3])
```
and I can't see how a `LocalVariable` can know that it's meant to be referenced with `&` with the information it has available.

Reply:
Perhaps the trace log is misleading? Our validation strictly requires the types of arguments to match, see https://github.com/gfx-rs/naga/blob/98252cf5d29790f85cc4397e0cf7a5b6cfaac614/src/valid/function.rs#L202
So the arguments passed into a call (of a function that expects pointers) are pointers, given that the validation succeeded.